### ISSUE TITLE
Error loading shared library /iframely/node_modules/iltorb/build/bindings/iltorb.node: Exec format 

### ISSUE THREAD

**Issue by bique14** [itteco/iframely#248]
When I run with docker from this [document](https://iframely.com/docs/host)

Its have an error like this and I cannot use iframely with docker.
<img width="986" alt="Screen Shot 2563-04-17 at 15 59 27" src="https://user-images.githubusercontent.com/24265604/79552100-c842fc80-80c4-11ea-9081-346fdfd55903.png">

**Comment by iparamonau** [comment]
Iframely's docker config and docs are community-contributed. Unfortunately, we don't be able to help with it.

**Comment by musically-ut** [comment]
It seems to be the case that the Docker version fails with Node version `12.18.1` while it works with node `12.13.1` and below.

I encountered this error, and the requests for _most_ URLs kept timing out. However, downgrading the node version fixed the issue.

**Comment by deshetti** [comment]
@bique14 Were you able to resolve this issue?

@musically-ut I changed the node version to `node:12.13.1-alpine` in the Dockefile and still getting the same error. I tried locally without Docker using Node version 12.13.1 and the latest LTS version of Node which is 12.19.0 and you are correct about timing out for most links with the latest version. But, changing the version to 12.13.1-alpine like I said still throws the same error. Wondering if they both relate to the same issues?

```
warn:    --minUptime not set. Defaulting to: 1000ms
warn:    --spinSleepTime not set. Your script will exit if it does not stay up for at least 1000ms
info:    Forever processing file: cluster.js
data:    cluster.js:31 - Using cache engine: node-cache
data:    cluster.js:31 - internal/modules/cjs/loader.js:807
data:    cluster.js:31 -   return process.dlopen(module, path.toNamespacedPath(filename));
data:    cluster.js:31 -                  ^
data:    cluster.js:31 - Error: Error loading shared library /iframely/node_modules/iltorb/build/bindings/iltorb.node: Exec format error
data:    cluster.js:31 -     at Object.Module._extensions..node (internal/modules/cjs/loader.js:807:18)
data:    cluster.js:31 -     at Module.load (internal/modules/cjs/loader.js:653:32)
data:    cluster.js:31 -     at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
data:    cluster.js:31 -     at Function.Module._load (internal/modules/cjs/loader.js:585:3)
data:    cluster.js:31 -     at Module.require (internal/modules/cjs/loader.js:692:17)
data:    cluster.js:31 -     at require (internal/modules/cjs/helpers.js:25:18)
data:    cluster.js:31 -     at Object.<anonymous> (/iframely/node_modules/iltorb/index.js:10:40)
data:    cluster.js:31 -     at Module._compile (internal/modules/cjs/loader.js:778:30)
data:    cluster.js:31 -     at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
data:    cluster.js:31 -     at Module.load (internal/modules/cjs/loader.js:653:32)
data:    cluster.js:31 - error: Forever detected script exited with code: 1
```

**Comment by deshetti** [comment]
@joltcan I followed the steps mentioned here: https://github.com/itteco/iframely/blob/master/docs/DOCKER-BUILD-INSTRUCTION.md

Please find my comment above for the error I am running into. Any help would be greatly appreciated.

**Comment by joltcan** [comment]
I have created a merge request from the Flattr branch that we are using inhouse, try https://github.com/flattr/iframely

**Comment by deshetti** [comment]
@joltcan Thanks for the quick response. I am getting the same error with your repository as well. I looked at the PR you opened and I have already made those changes in my local prior to testing.

**Comment by joltcan** [comment]
I wonder if this is due to your config? The updated makefile uses `config.js` instead of `config.local.js.SAMPLE` which doesn't work out of the box. I just built v1.5.0 and it seems to run locally at least, I have pushed it to https://hub.docker.com/repository/docker/jolt/iframely if you wanna test it.

**Comment by joltcan** [comment]
This is the config I build with btw:

```config.local.js

(function() {
    var config = {

        // Specify a path for custom plugins. Custom plugins will override core plugins.
        // CUSTOM_PLUGINS_PATH: __dirname + '/yourcustom-plugin-folder',

        DEBUG: true,
        RICH_LOG_ENABLED: true,

        // For embeds that require render, baseAppUrl will be used as the host.
        baseAppUrl: "http://localhost",
        relativeStaticUrl: "/r",

        // Or just skip built-in renders altogether
        SKIP_IFRAMELY_RENDERS: true,

        // For legacy reasons the response format of Iframely open-source is
        // different by default as it does not group the links array by rel.
        // In order to get the same grouped response as in Cloud API,
        // add `&group=true` to your request to change response per request
        // or set `GROUP_LINKS` in your config to `true` for a global change.
        GROUP_LINKS: true,

        // Number of maximum redirects to follow before aborting the page
        // request with `redirect loop` error.
        MAX_REDIRECTS: 4,

        SKIP_OEMBED_RE_LIST: [
            // /^https?:\/\/yourdomain\.com\//,
        ],

        /*
        // Used to pass parameters to the generate functions when creating HTML elements
        // disableSizeWrapper: Don't wrap element (iframe, video, etc) in a positioned div
        GENERATE_LINK_PARAMS: {
            disableSizeWrapper: true
        },
        */

        port: 8061, //can be overridden by PORT env var
        host: '0.0.0.0',    // Dockers beware. See https://github.com/itteco/iframely/issues/132#issuecomment-242991246
                            //can be overridden by HOST env var

        // Optional SSL cert, if you serve under HTTPS.
        /*
        ssl: {
            key: require('fs').readFileSync(__dirname + '/key.pem'),
            cert: require('fs').readFileSync(__dirname + '/cert.pem'),
            port: 443
        },
        */

        /*
        Supported cache engines:
        - no-cache - no caching will be used.
        - node-cache - good for debug, node memory will be used (https://github.com/tcs-de/nodecache).
        - redis - https://github.com/mranney/node_redis.
        - memcached - https://github.com/3rd-Eden/node-memcached
        */
        CACHE_ENGINE: 'node-cache',
        CACHE_TTL: 0, // In seconds. 
        // 0 = 'never expire' for memcached & node-cache to let cache engine decide itself when to evict the record
        // 0 = 'no cache' for redis. Use high enough (e.g. 365*24*60*60*1000) ttl for similar 'never expire' approach instead

        /*
        // Redis cache options.
        REDIS_OPTIONS: {
            host: '127.0.0.1',
            port: 6379
        },
        */

        /*
        // Memcached options. See https://github.com/3rd-Eden/node-memcached#server-locations
        MEMCACHED_OPTIONS: {
            locations: "127.0.0.1:11211"
        }
        */

        /*
        // Access-Control-Allow-Origin list.
        allowedOrigins: [
            "*",
            "http://another_domain.com"
        ],
        */

        /*
        // Uncomment to enable plugin testing framework.
        tests: {
            mongodb: 'mongodb://localhost:27017/iframely-tests',
            single_test_timeout: 10 * 1000,
            plugin_test_period: 2 * 60 * 60 * 1000,
            relaunch_script_period: 5 * 60 * 1000
        },
        */

        // If there's no response from remote server, the timeout will occur after
        RESPONSE_TIMEOUT: 5 * 1000, //ms

        /* From v1.4.0, Iframely supports HTTP/2 by default. Disable it, if you'd rather not.
           Alternatively, you can also disable per origin. See `proxy` option below.
        */
        // DISABLE_HTTP2: true,

        // Customize API calls to oembed endpoints.
        ADD_OEMBED_PARAMS: [{
            // Endpoint url regexp array.
            re: [/^http:\/\/api\.instagram\.com\/oembed/],
            // Custom get params object.
            params: {
                hidecaption: true
            }
        }, {
            re: [/^https:\/\/www\.facebook\.com\/plugins\/page\/oembed\.json/i],
            params: {
                show_posts: 0,
                show_facepile: 0,
                maxwidth: 600
            }
        }, {
            // match i=user or i=moment or i=timeline to configure these types invidually
            // see params spec at https://dev.twitter.com/web/embedded-timelines/oembed
            re: [/^https?:\/\/publish\.twitter\.com\/oembed\?i=user/i],
            params: {
                limit: 1,
                maxwidth: 600
            }
        /*
        }, {
            // Facebook https://developers.facebook.com/docs/plugins/oembed-endpoints
            re: [/^https:\/\/www\.facebook\.com\/plugins\/\w+\/oembed\.json/i],
            params: {
                // Skip script tag and fb-root div.
                omitscript: true
            }
        */
         }],

        /* Configure use of HTTP proxies as needed. 
           You don't have to specify all options per regex - just what you need to override
        */
        /*
        PROXY: [{
            re: [/^https?:\/\/www\.domain\.com/],
            proxy_server: 'http://1.2.3.4:8080',
            user_agent: 'CHANGE YOUR AGENT',
            headers: {
                // HTTP headers
                // Overrides previous params if overlapped.
            },
            request_options: {
                // Refer to: https://github.com/request/request
                // Overrides previous params if overlapped.
            },
            cache_ttl: 3600, // in seconds, cache response for 1 hour.
            disable_http2: true
        }],
        */

        // Customize API calls to 3rd parties. At the very least - configure required keys.
        providerOptions: {
            locale: "en_US",    // ISO 639-1 two-letter language code, e.g. en_CA or fr_CH. 
                                // Will be added as highest priotity in accept-language header with each request. 
                                // Plus is used in FB, YouTube and perhaps other plugins
            "twitter": {
                "max-width": 550,
                "min-width": 250,
                hide_media: false,
                hide_thread: false,
                omit_script: false,
                center: false,
                // dnt: true,                
                cache_ttl: 100 * 365 * 24 * 3600 // 100 Years.
            },
            readability: {
                enabled: false
                // allowPTagDescription: true  // to enable description fallback to first paragraph
            },
            images: {
                loadSize: false, // if true, will try an load first bytes of all images to get/confirm the sizes
                checkFavicon: false // if true, will verify all favicons
            },
            tumblr: {
                consumer_key: "INSERT YOUR VALUE"
                // media_only: true     // disables status embeds for images and videos - will return plain media
            },
            google: {
                // https://developers.google.com/maps/documentation/embed/guide#api_key
                //maps_key: "INSERT YOUR VALUE"
            },

            /*
            // Optional Camo Proxy to wrap all images: https://github.com/atmos/camo
            camoProxy: {
                camo_proxy_key: "INSERT YOUR VALUE",
                camo_proxy_host: "INSERT YOUR VALUE"
                // ssl_only: true // will only proxy non-ssl images
            },
            */

            // List of query parameters to add to YouTube and Vimeo frames
            // Start it with leading "?". Or omit alltogether for default values
            // API key is optional, youtube will work without it too.
            // It is probably the same API key you use for Google Maps.
            youtube: {
                // api_key: "INSERT YOUR VALUE",
                // parts: [ "snippet", "player" ], // list of fields you want to use in the request, in most cases you only need those two
                get_params: "?rel=0&showinfo=1"     // https://developers.google.com/youtube/player_parameters
            },
            vimeo: {
                get_params: "?byline=0&badge=0"     // https://developer.vimeo.com/player/embedding
            },

            /*
            soundcloud: {
                old_player: true // enables classic player
            },
            giphy: {
                media_only: true // disables branded player for gifs and returns just the image
            }
            */
            /*
            bandcamp: {
                get_params: '/size=large/bgcol=333333/linkcol=ffffff/artwork=small/transparent=true/',
                media: {
                    album: {
                        height: 472,
                        'max-width': 700
                    },
                    track: {
                        height: 120,
                        'max-width': 700
                    }
                }
            }
            */
        },

        // WHITELIST_WILDCARD, if present, will be added to whitelist as record for top level domain: "*"
        // with it, you can define what parsers do when they run accross unknown publisher.
        // If absent or empty, all generic media parsers will be disabled except for known domains
        // More about format: https://iframely.com/docs/qa-format

        /*
        WHITELIST_WILDCARD: {
              "twitter": {
                "player": "allow",
                "photo": "deny"
              },
              "oembed": {
                "video": "allow",
                "photo": "allow",
                "rich": "deny",
                "link": "deny"
              },
              "og": {
                "video": ["allow", "ssl", "responsive"]
              },
              "iframely": {
                "survey": "allow",
                "reader": "allow",
                "player": "allow",
                "image": "allow"
              },
              "html-meta": {
                "video": ["allow", "responsive"],
                "promo": "allow"
              }
        }
        */

        // Black-list any of the inappropriate domains. Iframely will return 417
        // At minimum, keep your localhosts blacklisted to avoid SSRF
        BLACKLIST_DOMAINS_RE: [
            /^https?:\/\/127\.0\.0\.1/i,
            /^https?:\/\/localhost/i,

            // And this is AWS metadata service
            // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html
            /^https?:\/\/169\.254\.169\.254/
        ]

        /** url for prerender service */
        //PRERENDER_URL: "https://domain/render?url=",
    };

    module.exports = config;
})();
```

**Comment by deshetti** [comment]
@joltcan I did change the `config.local.js.SAMPLE` in the makefile to `config.local.js` before trying.

In any case, I am using the exact same config.local.js file as yours from the above. You are correct, the dockerfile you uploaded to the hub is working as expected. Following is the first few lines when it starts on the console:

<img width="1196" alt="Screen Shot 2020-10-20 at 10 36 42 AM" src="https://user-images.githubusercontent.com/10763210/96609851-96536280-12c0-11eb-842d-6a7626c81497.png">

I am using the dockerfile I build with the same config file and get the following error:

<img width="1160" alt="Screen Shot 2020-10-20 at 10 36 23 AM" src="https://user-images.githubusercontent.com/10763210/96609933-af5c1380-12c0-11eb-9f56-1a4b046e0888.png">

I see that the issue is when it is trying to load the whitelist file: `[20-10-20 15:23:48]:32 Loading whitelist from https://iframely.com/qa/whitelist.json` that is when I get the error.

Would it be possible to share the `Dockerfile` and `entrypoint.sh` for me to compare?

**Different issue even with your image** Like @musically-ut pointed out in his comment, all the requests are timing out:
```
It seems to be the case that the Docker version fails with Node version 12.18.1 while it works with node 12.13.1 and below.

I encountered this error, and the requests for most URLs kept timing out. However, downgrading the node version fixed the issue.
```

**Comment by deshetti** [comment]
@joltcan I see that you might be using Node version: `12.18.3` instead of the latest `12.19.0` that it is coming by default when we use `FROM node:12-alpine` currently.

I picked up the above your latest image that you uploaded yesterday: https://hub.docker.com/layers/jolt/iframely/latest/images/sha256-3ac4fed145aca52cb31ec518bc437fc8fdccc77e0e7007f72b764b45e4fb4041?context=explore

The rest of the Dockerfile seems to be the same as I have it. The only thing I can think of is if there are any changes in the `entrypoint.sh` file. 

Thanks for your time, really appreciate it!

**Comment by joltcan** [comment]
All the files I use except the config is in the repo, which is just a fork of the iframely repo that I pull from when I build a new version, so the only difference is the the doc and my tiny Makefile changes, but it has nothing to do with the things inside the docker, just building the container. This is my [Dockerfile](https://github.com/flattr/iframely/blob/master/Dockerfile), so if alpine haven't tagged the latest correctly, well, then my images isn't updated either.

**Comment by joltcan** [comment]
Looks like I need to add `--pull` to `docker build` to be sure.

**Comment by joltcan** [comment]
Funny enough, that still didn't make a difference, it runs just great:

```
jolt@viggen:iframely [tag-v1.5.0]$ docker exec -it iframely cat /etc/alpine-release
3.11.6
 jolt@viggen:iframely [tag-v1.5.0]$ docker exec -it iframely node --version
v12.19.0
```

**Comment by deshetti** [comment]
@joltcan I was able to solve the issue by adding node_modules to .dockerignore. Found it the hard way, where it was working for me when I login to the container, delete the node_modules and restart the container. 

I opened a new PR with all your changes and this fix. Also, included node version to be `12.13.1` in the dockerfile now to work around the timeout issue.

Thanks for all your time so far.

**Comment by iparamonau** [comment]
Thanks @joltcan @deshetti for contributions! It's really appreciated.  We'll review till the end of the week and merge.
